### ISSUE TITLE
how to set segment_id in pretrained_transformer model

### ISSUE THREAD

**Issue by wj-Mcat** [allenai/allennlp#4590]
I train my sequence tagging model, here is my configuration file.
```json
{
  "dataset_reader": {
    "type": "ccks-sequence_tagging",
    "token_indexers": {
      "tokens": {
        "type": "pretrained_transformer_mismatched",
        "model_name": "bert-base-chinese"
      }
    }
  },
  "train_data_path": "./data/train_base_phrase_clear.txt",
  "data_loader": {
    "batch_size": 4,
    "shuffle": true
  },
  "model": {
    "type": "crf_tagger",
    "text_field_embedder": {
      "token_embedders": {
        "tokens": {
          "type": "pretrained_transformer_mismatched",
          "model_name": "bert-base-chinese"
        }
      }
    },
    "encoder": {
      "type": "pass_through",
      "input_dim":768
    },
    "calculate_span_f1": true,
    "label_encoding": "BIO"
  },
  "trainer": {
    "optimizer": "adam",
    "patience": 10,
    "cuda_device": 2,
    "num_epochs": 2
  }
}
```
And the data in `train_base_phrase_clear.txt` file is like: `word###label word###label [SEP]###O sentence-question-word###O sentence-question-word###O`.

My main problem is that the result from token_embedders doesn't contains type_ids info, which is all 0 value. I have set `[SEP]` word label in my trainning file, So I think there should be different type_ids. 

So, I finnaly check that there is no type_id info from `SequenceTaggingDatasetReader` class. Then I custom this reader and add init `PretrainedTransformerTokenizer` instance to this reader, and add type_ids to Instance. B-U-T, when the data flow from dataset_reader to crf_tagger module, there is no type_id info. 

This make me confused. where does `type_id` info go away ? And how to tokenize sentence with segment_id information in allennlp ? 

This one is urgent. Waiting for your reply! Thanks for your later help.

**Comment by matt-gardner** [comment]
@dirkgr can correct me if I'm wrong, but I believe the right thing to do is to call `tokenizer.add_special_tokens(before_sep, after_sep)` on your `PretrainedTransformerTokenizer`.  You'll have to manually split on the `[SEP]` token.  If you just call `tokenizer.tokenize()`, we call huggingface's `encode_plus`, which does not detect `[SEP]` tokens to set type ids.  We used to have code that detected `[SEP]` tokens and incremented type ids, but that got lost somewhere in the upgrade from our `BertIndexer` to what we have now.

**Comment by dirkgr** [comment]
If you don't have pre-tokenized text, then calling `add_special_tokens()` like @matt-gardner said is correct. But that will not do the right thing when your text is already tokenized.

In fact, when you use the `PretrainedTransformerMismatchedIndexer`, we ignore the type ids that the reader might have set on the tokens, so there is no way to get the type ids to the right place. I think we should fix [this line](https://github.com/allenai/allennlp/blob/master/allennlp/data/token_indexers/pretrained_transformer_mismatched_indexer.py#L74) so it uses the type id from the token instead of from the word pieces. Then your reader can set type ids, and they will be respected by the mismatched indexer. @wj-Mcat, do you think that will work?

**Comment by dirkgr** [comment]
I have looked at this a little bit closer, and found it's not that easy. The mismatched indexer also adds special tokens, but it has no way of knowing anything about sequence pairs. So even if we create a way to pass token type ids through to the mismatched indexer, we can't guarantee that it'll create the correct special tokens. @wj-Mcat is "solving" the problem by setting the middle special token himself, but relying on the mismatched indexer to place the special token at the front and back of the sequence. That's very brittle, and only happens to work for a few models, not in general.

I don't think you can use the mismatched indexer for this. It's designed to make old-school datasets work with transformers. If you already have transformer-specific input data, it won't be very nice to use.

As a workaround, I recommend you use the regular pretrained transformer indexer and embedder, and  keep track of the offsets yourself, as part of the instances that the reader produces. You can probably steal a lot of the code from mismatched indexer/embedder, and you can use `intra_word_tokenize` of course.

**Comment by wj-Mcat** [comment]
@dirkgr  I need to tokenize the sentence at character level, but `PretrainedTransformerMismatchedIndexer ` will tokenize the `2019.08.23` data to ['2019', '.', '08', '23'] data. And sometimes, it will ads `##` prefix to the token text. These features will make me hard to set character-level tag.  So, I tokenize the text before training.

**Comment by dirkgr** [comment]
To use the pretrained transformer, you have to tokenize exactly the way it was tokenized when BERT was trained. Otherwise the pre-trained BERT weights don't work. That is what `PretrainedTransformerTokenizer` does, and that is what `PretrainedTransformerIndexer` expects. This is why the date gets split up like that, and that's why you see the `"##"` prefix.

I understand that your tags use a different tokenization. You said "character-level", but I think you mean "word-level"? Either way, I think you need to tokenize the input text with `PretrainedTransformerTokenizer`, maybe with `intra_word_tokenize`, and keep track of the mapping between BERT word pieces and words yourself. That way you can set token IDs, add special tokens, and do other stuff as you need to. We have some functions that help with that.

It is strange to see that your input data on the one hand uses a tokenization that doesn't match with BERT tokenization, but on the other hand it uses [SEP] tokens that are specific to BERT. It does not make a lot of sense to have data this way. This is why the "mismatched" group of classes doesn't help you.

**Comment by wj-Mcat** [comment]
When I review the transformers source code, I found that it support segment feature is supported. 

```python
    @add_end_docstrings(ENCODE_KWARGS_DOCSTRING, ENCODE_PLUS_ADDITIONAL_KWARGS_DOCSTRING)
    def encode_plus(
        self,
        text: Union[TextInput, PreTokenizedInput, EncodedInput],
        text_pair: Optional[Union[TextInput, PreTokenizedInput, EncodedInput]] = None,
        add_special_tokens: bool = True,
        padding: Union[bool, str] = False,
        truncation: Union[bool, str] = False,
        max_length: Optional[int] = None,
        stride: int = 0,
        is_pretokenized: bool = False,
        pad_to_multiple_of: Optional[int] = None,
        return_tensors: Optional[Union[str, TensorType]] = None,
        return_token_type_ids: Optional[bool] = None,
        return_attention_mask: Optional[bool] = None,
        return_overflowing_tokens: bool = False,
        return_special_tokens_mask: bool = False,
        return_offsets_mapping: bool = False,
        return_length: bool = False,
        verbose: bool = True,
        **kwargs
    ) -> BatchEncoding:
```
I think `text_pair` is the mainly supported feature for segment_id, which is the `token_type_id` in allennlp. 

But you haven't make the sentence pair tokenized by pair mode.  Am I right ?

**Comment by dirkgr** [comment]
You could also do this, yes.

**Comment by VisualJoyce** [comment]
I am facing a similar issue on version 2.0.0

I have to do the following:
```
        tokenized_a = self._tokenizer.tokenize(example.text_a)
        tokenized_b = self._tokenizer.tokenize(example.text_b)
        tokenized_source = self._token_indexers['bert']._allennlp_tokenizer.add_special_tokens(tokenized_a,
                              tokenized_b)[1:-1]
        source_field = TextField(tokenized_source, self._token_indexers)
```
Trimming the special tokens added by `_allennlp_tokenizer` on both ends and  then construct the `TextField`. Otherwise, the token indexer will add one more `[CLS]` and `[SEP]`.

**Comment by dirkgr** [comment]
Did you initialize `_allennlp_tokenizer` with `add_special_tokens=False`?

**Comment by VisualJoyce** [comment]
The special tokens are added by the inner tokenizer of `PretrainedTransformerIndexer`.

https://github.com/allenai/allennlp/blob/f877fdc30d18178b88c335fbd92722fb77c42d93/allennlp/data/token_indexers/pretrained_transformer_indexer.py#L57-L60

By default, it is initialized with `add_special_tokens=True`.

**Comment by VisualJoyce** [comment]
I am using a spacy tokenizer, the config is:
```
  "dataset_reader": {
    "type": "glue",
    "task_name": TASK_NAME,
    'tokenizer': {
      "type": "spacy",
      "pos_tags": POS_TAGS,
      "language": LANGUAGE + "_core_web_sm"
    },
    "token_indexers": {
      "bert": {
        "type": "pretrained_transformer_mismatched",
        "max_length": 512,
        "model_name": MODEL_NAME
      }
    }
  },
```

**Comment by VisualJoyce** [comment]
Now I switched to the head of the main branch and would like to offer more details:
```python
        tokenized_a = self._tokenizer.tokenize(example.text_a)
        if example.text_b:
            tokenized_b = self._tokenizer.tokenize(example.text_b)
            if self._is_transformer_tokenizer:
                tokenized_source = self._token_indexers['bert']._allennlp_tokenizer.add_special_tokens(
                    tokenized_a[1:-1],
                    tokenized_b[1:-1])
            else:
                tokenized_source = self._token_indexers['bert']._allennlp_tokenizer.add_special_tokens(tokenized_a,
                                                                                                       tokenized_b)[1:-1]

            source_field = TextField(tokenized_source)
        else:
            source_field = TextField(tokenized_a)
```

In this way, I can ensure the tokenized source contains the correct number of [CLS] and [SEP].

However, I am facing the issue that `type_ids` are all zeros, which means these `type_ids` might have been changed during token indexing.

**Comment by VisualJoyce** [comment]
My guess for the issue above is tokens lost their `type_id` in the `intra_word_tokenize`.

https://github.com/allenai/allennlp/blob/90915800d645c0e5a17e01e56f1fd4d02c5f4e43/allennlp/data/token_indexers/pretrained_transformer_mismatched_indexer.py#L74-L80

**Comment by dirkgr** [comment]
The `*mismatched*` stuff does not support type ids. It would be a little difficult to do, because it operates on strings, not tokens, and strings don't have type IDs. You could probably fix this in `PretrainedTransformerMismatchedIndexer.tokens_to_indices`, because that function receives `Token`. We could put some code in there that preserves the type ids from the input. But usually those tokens will come from Spacy, which doesn't produce type ids anyways. Are you making sure in your reader that the type ids are properly set? If so, I would recommend changing that function to preserve type ids, and please submit a PR with that change :-)
